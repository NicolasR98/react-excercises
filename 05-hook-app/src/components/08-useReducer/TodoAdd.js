import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
	const [{ description }, handleInputChange, resetForm] = useForm({
		description: '',
	});

	const handleSubmit = (e) => {
		e.preventDefault();

		if (description.trim().length <= 1) return;

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false,
		};

		handleAddTodo(newTodo);
		resetForm();
	};

	return (
		<>
			<h4>Add ToDo</h4>
			<hr />

			<form onSubmit={handleSubmit}>
				<input
					onChange={handleInputChange}
					value={description}
					className="form-control"
					type="text"
					name="description"
					placeholder="Do ..."
					autoComplete="off"
				/>
				<button type="submit" className="btn btn-success col-12 mt-3">
					Add ToDo
				</button>
			</form>
		</>
	);
};
