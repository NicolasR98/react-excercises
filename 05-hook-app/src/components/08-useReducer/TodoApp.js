import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hooks/useForm';
import { todoReducer } from './todoReducer';

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {
	const [todos, dispatch] = useReducer(todoReducer, [], init);

	const [{ description }, handleInputChange, resetForm] = useForm({
		description: '',
	});

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));
	}, [todos]);

	const handleDelete = (todoId) => {
		const action = {
			type: 'delete',
			payload: todoId,
		};
		dispatch(action);
	};

	const handleToggle = (todoId) => {
		dispatch({
			type: 'toggle',
			payload: todoId,
		});
	};

	const handleSubmit = (e) => {
		e.preventDefault();

		if (description.trim().length <= 1) return;

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false,
		};

		const action = {
			type: 'add',
			payload: newTodo,
		};

		dispatch(action);
		resetForm();
	};

	return (
		<div>
			<h1>TodoApp ({todos.length})</h1>
			<hr />

			<div className="row">
				<div className="col-7">
					<ul className="list-group list-group-flush">
						{todos.map((todo, index) => (
							<li key={todo?.id} className="list-group-item">
								<p
									className={`${todo.done && 'complete'}`}
									onClick={() => handleToggle(todo.id)}
								>
									{index + 1}. {todo?.desc}
								</p>
								<button
									onClick={() => handleDelete(todo.id)}
									className="btn btn-danger btn-sm"
								>
									Delete
								</button>
							</li>
						))}
					</ul>
				</div>
				<div className="col-5">
					<h4>Add ToDo</h4>
					<hr />

					<form onSubmit={handleSubmit}>
						<input
							onChange={handleInputChange}
							value={description}
							className="form-control"
							type="text"
							name="description"
							placeholder="Do ..."
							autoComplete="off"
						/>
						<button type="submit" className="btn btn-success col-12 mt-3">
							Add ToDo
						</button>
					</form>
				</div>
			</div>
		</div>
	);
};
