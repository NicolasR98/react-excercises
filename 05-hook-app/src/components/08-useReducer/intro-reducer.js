const initialState = [
	{
		id: 1,
		todo: 'buy groceries',
		done: false,
	},
];

const todoReducer = (state = initialState, action) => {
	if (action?.type === 'add') {
		return [...state, action.payload];
	}

	return state;
};

let todos = todoReducer();

const newTodo = {
	id: 1,
	todo: 'buy groceries',
	done: true,
};

const addTodoAction = {
	type: 'add',
	payload: newTodo,
};

todos = todoReducer(todos, addTodoAction);

console.log(todos);
