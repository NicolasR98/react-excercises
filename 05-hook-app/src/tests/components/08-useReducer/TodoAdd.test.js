import React from 'react';
import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('Tests on <TodoAdd />', () => {
	const handleAddTodo = jest.fn();

	let wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

	test('should display correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should not call `handleAddTodo`', () => {
		const formSubmit = wrapper.find('form').prop('onSubmit');

		formSubmit({ preventDefault() { } });

		expect(handleAddTodo).not.toHaveBeenCalled();
	});

	test('should call `handleAddTodo`', () => {
		const value = 'Learn React';

		wrapper.find('input').simulate('change', {
			target: {
				value,
				name: 'description'
			}
		});

		const formSubmit = wrapper.find('form').prop('onSubmit');
		formSubmit({ preventDefault() { } });

		expect(handleAddTodo).toHaveBeenCalledTimes(1);
		expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
		expect(handleAddTodo).toHaveBeenCalledWith({
			id: expect.any(Number),
			desc: value,
			done: false
		});

		expect(wrapper.find('input').prop('value')).toBe('');
	});

});
