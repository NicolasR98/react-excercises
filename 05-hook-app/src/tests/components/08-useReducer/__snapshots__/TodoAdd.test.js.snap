// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TodoAdd /> should display correctly 1`] = `
<Fragment>
  <h4>
    Add ToDo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Do ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-success col-12 mt-3"
      type="submit"
    >
      Add ToDo
    </button>
  </form>
</Fragment>
`;
